### PR TITLE
[WIP] Cluster resource limits

### LINKED ISSUE #249: Support resource limits for an individual cluster
Previously we supported configuring limits *per user*. These included:
- Max clusters per user
- Max cores per user
- Max memory per user

These were dropped in the refactor, as most backends have a native way of managing resource (cpu/memory) limits. 

- Kubernetes has [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
- YARN, PBS, SLURM, etc... have queues

Supporting them in the gateway also complicated the implementation - most decisions are made at the *cluster level*, but limits on cores/memory were handled at the *user level*.

However, using the backend's native method does impose some restrictions on gateway configuration that may not be desirable for certain admins/users. For example, in Kubernetes, `ResourceQuota` objects are associated with a namespace. If you want to have limits per user, each user would need their own namespace.

I don't think going back to the old model is tenable, as I'd prefer to keep all cluster decisions local to information about a single cluster (rather than global information about multiple clusters). But, we can support some simpler limits natively fairly easily. These would include:

- Max workers per cluster
- Max cores per cluster
- Max memory per cluster

These limits would exist on the `ClusterConfig` object, and could thus be influenced by [user-specific cluster options](https://gateway.dask.org/cluster-options.html). So you could change limits on cluster size per user (or per group, etc...).

### PR DESCRIPTION
**Pull request by jcrist**
This adds resource limits *per cluster*. Currently we support:

- Max cores per cluster
- Max memory per cluster
- Max workers per cluster

At runtime these are normalized into a single `cluster_max_workers`
field, which is used to check incoming scale and adapt requests. If a
request from a user exceeds the limit, it is trimmed to be within
bounds, and a warning is raised in the user's terminal notifying them of
the limit.

Still needs tests and docs.

Fixes #249.

### CODE REVIEW

File: dask-gateway-server/dask_gateway_server/backends/base.py [hunk 1]
```diff
@@ -320,6 +346,68 @@ class ClusterConfig(Configurable):
...
+
+        Set to ``None`` for no limit (default).
+        """,
+        allow_none=True,
+        min=0,
+        config=True,
+    )
+
+    @default("cluster_max_workers")
+    def _default_cluster_max_workers(self):
```
Comment (jcrist) [inline dask-gateway-server/dask_gateway_server/backends/base.py:394]: If `cluster_max_workers` is explicitly set, then `cluster_max_cores`/`cluster_max_memory` are ignored, even if those restrictions are stricter. I'm waffling on whether that's the right behavior. Right now I think that's confusing, and we should take the min of the 3.

File: dask-gateway-server/dask_gateway_server/routes.py [hunk 2]
```diff
@@ -226,14 +236,27 @@ async def adapt_cluster(request):
     maximum = msg.get("maximum", None)
     active = msg.get("active", True)
 
+    max_workers = cluster.config.get("cluster_max_workers")
+    resp_msg = None
+    if max_workers is not None:
+        if maximum is None:
+            maximum = max_workers
```
Comment (jcrist) [inline dask-gateway-server/dask_gateway_server/routes.py:243]: Just realized we'll also want to apply this to `minimum` as well.

Comment (jcrist) [reply dask-gateway-server/dask_gateway_server/routes.py:243]: this whole `adapt` call looks to be missing some input validation period.

File: dask-gateway-server/dask_gateway_server/backends/base.py [hunk 3]
```diff
@@ -320,6 +346,68 @@ class ClusterConfig(Configurable):
...
+        help="""
+        The maximum amount of memory (in bytes) available to this cluster.
+        Allows the following suffixes:
+
+        - K -> Kibibytes
+        - M -> Mebibytes
+        - G -> Gibibytes
+        - T -> Tebibytes
+
+        Set to ``None`` for no limit (default).
```
Comment (ericdill) [inline dask-gateway-server/dask_gateway_server/backends/base.py:360]: May want to add "Limits may still be imposed by `cluster_max_cores` and `cluster_max_workers`. Similar for the other two

File: dask-gateway-server/dask_gateway_server/backends/base.py (hunk 1 above)
Comment (ericdill) [reply dask-gateway-server/dask_gateway_server/backends/base.py:394]: I'm +1 on taking the min of the 3. That seems more intuitive

**Pull request merged**